pq-crypto/sike_r3/fpx.c

```diff
@@ -0,0 +1,475 @@
+/********************************************************************************************
+* Supersingular Isogeny Key Encapsulation Library
+*
+* Abstract: core functions over GF(p) and GF(p^2)
+*********************************************************************************************/
+
+#include <string.h>
+#include "sikep434r3.h"
+#include "fp.h"
+#include "fpx.h"
+
+static void fpmul_mont(const felm_t ma, const felm_t mb, felm_t mc);
+static void to_mont(const felm_t a, felm_t mc);
+static void from_mont(const felm_t ma, felm_t c);
+static void fpsqr_mont(const felm_t ma, felm_t mc);
+static unsigned int mp_sub(const digit_t* a, const digit_t* b, digit_t* c, const unsigned int nwords);
+static void fpinv_chain_mont(felm_t a);
+static void fpinv_mont(felm_t a);
+static void to_fp2mont(const f2elm_t *a, f2elm_t *mc);
+static void from_fp2mont(const f2elm_t *ma, f2elm_t *c);
+
+/* Compare two byte arrays in constant time.
+ * Returns 0 if the byte arrays are equal, -1 otherwise. */
+int8_t ct_compare(const uint8_t *a, const uint8_t *b, unsigned int len)
+{
+    uint8_t r = 0;
+
+    for (unsigned int i = 0; i < len; i++) {
+        r |= a[i] ^ b[i];
+    }
+
+    return (-(int8_t)r) >> (8*sizeof(uint8_t)-1);
+}
+
+/* Conditional move in constant time.
+ * If selector = -1 then load r with a, else if selector = 0 then keep r. */
+void ct_cmov(uint8_t *r, const uint8_t *a, unsigned int len, int8_t selector)
+{
+    for (unsigned int i = 0; i < len; i++) {
+        r[i] ^= selector & (a[i] ^ r[i]);
+    }
+}
```

Comment:
I've seen at least 3 distinct implementations now of both of these operations. We should be using the functions in `s2n_safety.c`

Reply:
I've removed these functions and replace them with s2n's implementations.